components/director/internal/runtimemapping/token_verifier.go

```diff
@@ -0,0 +1,263 @@
+	flag   sync.Mutex
+}
+
+func NewJWKsCache(logger Logger, fetch KeyGetter) *jwksCache {
+	return &jwksCache{
+		logger: logger,
+		fetch:  fetch,
+		cache:  make(map[string]jwkCacheEntry),
+	}
+}
+
+func (c *jwksCache) GetKey(token *jwt.Token) (interface{}, error) {
+	if token == nil {
+		return nil, errors.New("token cannot be nil")
+	}
+
+	keyID, err := getKeyID(*token)
+	if err != nil {
+		return nil, errors.Wrap(err, "while getting the key ID")
+	}
+
+	c.flag.Lock()
+	cachedKey, exists := c.cache[keyID]
+	c.flag.Unlock()
+
+	if !exists || cachedKey.IsExpired() {
+		key, err := c.fetch.GetKey(token)
+		if err != nil {
+			return nil, errors.Wrapf(err, "while getting the key with ID [kid=%s]", keyID)
+		}
+
+		c.logger.Info(fmt.Sprintf("adding key %s to cache", keyID))
+
+		c.flag.Lock()
+		c.cache[keyID] = jwkCacheEntry{
+			key:      key,
+			cachedAt: time.Now(),
+		}
+		c.flag.Unlock()
+
+		return key, nil
+	}
+
+	c.logger.Info(fmt.Sprintf("using key %s from cache", keyID))
+
+	return cachedKey.key, nil
+}
+
+func (c *jwksCache) Cleanup() {
+	for keyID := range c.cache {
+		if !c.cache[keyID].IsExpired() {
+			continue
+		}
+
+		c.logger.Info(fmt.Sprintf("removing key %s from cache", keyID))
+
+		c.flag.Lock()
+		delete(c.cache, keyID)
+		c.flag.Unlock()
+	}
+}
+
+type jwksFetch struct {
+	logger Logger
+}
+
+func NewJWKsFetch(logger Logger) *jwksFetch {
+	return &jwksFetch{
+		logger: logger,
+	}
+}
+
+func (f *jwksFetch) GetKey(token *jwt.Token) (interface{}, error) {
+	if token == nil {
+		return nil, errors.New("token cannot be nil")
+	}
+
+	jwksURI, err := f.getJWKsURI(*token)
+	if err != nil {
+		return nil, errors.Wrap(err, "while getting the JWKs URI")
+	}
+
+	jwksSet, err := jwk.Fetch(jwksURI)
+	if err != nil {
+		return nil, errors.Wrap(err, "while fetching JWKs")
+	}
+
+	keyID, err := getKeyID(*token)
+	if err != nil {
+		return nil, errors.Wrap(err, "while getting the key ID")
+	}
+
+	keys := jwksSet.LookupKeyID(keyID)
+	for _, key := range keys {
+		if key.Algorithm() == token.Method.Alg() {
+			return key.Materialize()
+		}
+	}
+
+	return nil, errors.New("unable to find a proper key")
+}
+
+func (f *jwksFetch) getJWKsURI(token jwt.Token) (string, error) {
+	discoveryURL, err := getDiscoveryURL(token)
+	if err != nil {
+		return "", errors.Wrap(err, "while getting the discovery URL")
+	}
+
+	res, err := http.Get(discoveryURL)
+	if err != nil {
+		return "", errors.Wrap(err, "while getting the configuration discovery")
+	}
+	defer func() {
+		if err := res.Body.Close(); err != nil {
+			f.logger.Error(err)
+		}
+	}()
+
+	var config map[string]interface{}
+	if err = json.NewDecoder(res.Body).Decode(&config); err != nil {
+		return "", errors.Wrap(err, "while decoding the configuration discovery response")
+	}
+
+	jwksURI, ok := config[jwksURIKey].(string)
+	if !ok {
+		return "", errors.New("unable to cast the JWKs URI to a string")
+	}
+
+	return jwksURI, nil
+}
+
+func getDiscoveryURL(token jwt.Token) (string, error) {
```

Comment:
`getDiscoveryURL` is used only by `jwksFetch` maybe convert it to the `jwksFetch` method, the same for other functions.

Reply:
done